src/scatter_buildblock/ScatterEstimationByBin.cxx

```diff
@@ -11,447 +11,1274 @@
+                    read_from_file<DiscretisedDensity<3, float> >(this->mask_atten_image_filename);
+        }
+
+        if (this->tail_mask_par_filename.size() == 0)
+        {
+            warning("Please define a filename for tails mask. Abord.");
+            return true;
+        }
+    }
+
+    return false;
 }
 
-void
+Succeeded
 ScatterEstimationByBin::
-set_activity_image(const std::string& filename)  
+set_up()
 {
-  this->activity_image_filename=filename;
-  this->activity_image_sptr= 
-    read_from_file<DiscretisedDensity<3,float> >(filename);
-  if (is_null_ptr(this->activity_image_sptr))
+    if (this->run_debug_mode)
     {
-      error(boost::format("Error reading activity image %s") %
-            this->activity_image_filename);
+        info("Debugging mode is activated.");
+        this->export_2d_projdata = true;
+        this->export_scatter_estimates_of_each_iteration = true;
     }
-  this->set_activity_image_sptr(this->activity_image_sptr);
-}
 
+    if (is_null_ptr(this->input_projdata_sptr))
+    {
+        warning(boost::format("Input projdata file %1% not found. Abord.") %this->input_projdata_filename );
+        return Succeeded::no;
+    }
 
-void
-ScatterEstimationByBin::
-set_density_image_sptr(const shared_ptr<DiscretisedDensity<3,float> >& new_sptr)
-{
-  this->density_image_sptr=new_sptr;
-  this->remove_cache_for_integrals_over_attenuation();
+    // Load InputProjData and calculate the SSRB
+    if (this->input_projdata_sptr->get_max_segment_num() > 0 )
+    {
+        this->proj_data_info_2d_sptr.reset(
+                    dynamic_cast<ProjDataInfoCylindricalNoArcCorr* >
+                    (SSRB(*this->input_projdata_sptr->get_proj_data_info_ptr(),
+                          this->input_projdata_sptr->get_max_segment_num())));
+
+        if (this->export_2d_projdata)
+        {
+            size_t lastindex = this->input_projdata_filename.find_last_of(".");
+            std::string rawname = this->input_projdata_filename.substr(0, lastindex);
+            std::string out_filename = rawname + "_2d.hs";
+
+            this->input_projdata_2d_sptr.reset(new ProjDataInterfile(this->input_projdata_sptr->get_exam_info_sptr(),
+                                                                     this->proj_data_info_2d_sptr,
+                                                                     out_filename,
+                                                                     std::ios::in | std::ios::out | std::ios::trunc));
+        }
+        else
+            this->input_projdata_2d_sptr.reset(new ProjDataInMemory(this->input_projdata_sptr->get_exam_info_sptr(),
+                                                                    this->proj_data_info_2d_sptr));
+
+        SSRB(*this->input_projdata_2d_sptr,
+             *input_projdata_sptr,false);
+    }
+    else
+    {
+        warning(boost::format("The input data %1% are not 3D. Abord.") %this->input_projdata_filename );
```

Comment:
The proper would be if the user input 2D data then the final up-sampling be deactivated and provide an method to do it in later time ( if wanted ).